File: binding/src/network/libqhostaddress.cpp
#include <QHostAddress>
#include <QIPv6Address>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qhostaddress.h>
#include "libqhostaddress.h"
#include "libqhostaddress.hxx"

QIPv6Address* QIPv6Address_new() {
	 return new QIPv6Address();
}

QIPv6Address* QIPv6Address_new2(const QIPv6Address* param1) {
	 return new QIPv6Address(*param1);
}

unsigned char* QIPv6Address_OperatorSubscript(QIPv6Address* self, int index) {
	return self->operator[](index);
}

unsigned char QIPv6Address_OperatorSubscript2(const QIPv6Address* self, int index) {
	return self->operator[](index);
}

void QIPv6Address_OperatorAssign(QIPv6Address* self, const QIPv6Address* param1) {
	self->operator=(*param1);
}

void QIPv6Address_Delete(QIPv6Address* self) {
    delete self;
}

QHostAddress* QHostAddress_new() {
	 return new QHostAddress();
}

QHostAddress* QHostAddress_new2(unsigned int ip4Addr) {
	 return new QHostAddress(ip4Addr);
}

QHostAddress* QHostAddress_new3(const unsigned char* ip6Addr) {
	 return new QHostAddress(ip6Addr);
}

QHostAddress* QHostAddress_new4(const QIPv6Address* ip6Addr) {
	 return new QHostAddress(*ip6Addr);
}

QHostAddress* QHostAddress_new5(const libqt_string address) {
	 return new QHostAddress(QString::fromUtf8(address.data, address.len));
}

QHostAddress* QHostAddress_new6(const QHostAddress* copyVal) {
	 return new QHostAddress(*copyVal);
}

QHostAddress* QHostAddress_new7(int address) {
	 return new QHostAddress(static_cast<QHostAddress::SpecialAddress>(address));
}

void QHostAddress_OperatorAssign(QHostAddress* self, const QHostAddress* other) {
	self->operator=(*other);
}

void QHostAddress_OperatorAssign2(QHostAddress* self, int address) {
	self->operator=(static_cast<QHostAddress::SpecialAddress>(address));
}

void QHostAddress_Swap(QHostAddress* self, QHostAddress* other) {
	self->swap(*other);
}

void QHostAddress_SetAddress(QHostAddress* self, unsigned int ip4Addr) {
	self->setAddress(ip4Addr);
}

void QHostAddress_SetAddress2(QHostAddress* self, const unsigned char* ip6Addr) {
	self->setAddress(ip6Addr);
}

void QHostAddress_SetAddress3(QHostAddress* self, const QIPv6Address* ip6Addr) {
	self->setAddress(*ip6Addr);
}

bool QHostAddress_SetAddress5(QHostAddress* self, const libqt_string address) {
	return self->setAddress(QString::fromUtf8(address.data, address.len));
}

void QHostAddress_SetAddress6(QHostAddress* self, int address) {
	self->setAddress(static_cast<QHostAddress::SpecialAddress>(address));
}

int QHostAddress_Protocol(const QHostAddress* self) {
	return self->protocol();
}

unsigned int QHostAddress_ToIPv4Address(const QHostAddress* self) {
	return self->toIPv4Address();
}

QIPv6Address* QHostAddress_ToIPv6Address(const QHostAddress* self) {
	return new QIPv6Address(self->toIPv6Address());
}

libqt_string QHostAddress_ToString(const QHostAddress* self) {
	QString _ret = self->toString();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QHostAddress_ScopeId(const QHostAddress* self) {
	QString _ret = self->scopeId();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QHostAddress_SetScopeId(QHostAddress* self, const libqt_string id) {
	self->setScopeId(QString::fromUtf8(id.data, id.len));
}

bool QHostAddress_IsEqual(const QHostAddress* self, const QHostAddress* address) {
	return self->isEqual(*address);
}

bool QHostAddress_OperatorEqual(const QHostAddress* self, const QHostAddress* address) {
	return self->operator==(*address);
}

bool QHostAddress_OperatorEqual2(const QHostAddress* self, int address) {
	return self->operator==(static_cast<QHostAddress::SpecialAddress>(address));
}

bool QHostAddress_OperatorNotEqual(const QHostAddress* self, const QHostAddress* address) {
	return self->operator!=(*address);
}

bool QHostAddress_OperatorNotEqual2(const QHostAddress* self, int address) {
	return self->operator!=(static_cast<QHostAddress::SpecialAddress>(address));
}

bool QHostAddress_IsNull(const QHostAddress* self) {
	return self->isNull();
}

void QHostAddress_Clear(QHostAddress* self) {
	self->clear();
}

bool QHostAddress_IsInSubnet(const QHostAddress* self, const QHostAddress* subnet, int netmask) {
	return self->isInSubnet(*subnet, netmask);
}

bool QHostAddress_IsLoopback(const QHostAddress* self) {
	return self->isLoopback();
}

bool QHostAddress_IsGlobal(const QHostAddress* self) {
	return self->isGlobal();
}

bool QHostAddress_IsLinkLocal(const QHostAddress* self) {
	return self->isLinkLocal();
}

bool QHostAddress_IsSiteLocal(const QHostAddress* self) {
	return self->isSiteLocal();
}

bool QHostAddress_IsUniqueLocalUnicast(const QHostAddress* self) {
	return self->isUniqueLocalUnicast();
}

bool QHostAddress_IsMulticast(const QHostAddress* self) {
	return self->isMulticast();
}

bool QHostAddress_IsBroadcast(const QHostAddress* self) {
	return self->isBroadcast();
}

bool QHostAddress_IsPrivateUse(const QHostAddress* self) {
	return self->isPrivateUse();
}

libqt_pair QHostAddress_ParseSubnet(const libqt_string subnet) {
	return QHostAddress::parseSubnet(QString::fromUtf8(subnet.data, subnet.len));
}

unsigned int QHostAddress_ToIPv4Address1(const QHostAddress* self, bool* ok) {
	return self->toIPv4Address(ok);
}

bool QHostAddress_IsEqual2(const QHostAddress* self, const QHostAddress* address, int mode) {
	return self->isEqual(*address, static_cast<QFlags<QHostAddress::ConversionModeFlag>>(mode));
}

void QHostAddress_Delete(QHostAddress* self) {
    delete self;
}

File: binding/include/network/libqhostaddress.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qipv6address.html
pub const qipv6address = struct {

    /// New constructs a new QIPv6Address object.
    pub fn New() QtC.QIPv6Address {
        return qtc.QIPv6Address_new();
    }


    /// New2 constructs a new QIPv6Address object.
    pub fn New2(param1: ?*anyopaque) QtC.QIPv6Address {
        return qtc.QIPv6Address_new2(@ptrCast(param1));
    }


    pub fn OperatorSubscript(self: ?*anyopaque, index: i32) u8 {
        return @ptrCast(qtc.QIPv6Address_OperatorSubscript(@ptrCast(self), index));
    }

    pub fn OperatorSubscript2(self: ?*anyopaque, index: i32) u8 {
        return qtc.QIPv6Address_OperatorSubscript2(@ptrCast(self), index);
    }

    pub fn OperatorAssign(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QIPv6Address_OperatorAssign(@ptrCast(self), @ptrCast(param1));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QIPv6Address_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qhostaddress.html
pub const qhostaddress = struct {

    /// New constructs a new QHostAddress object.
    pub fn New() QtC.QHostAddress {
        return qtc.QHostAddress_new();
    }


    /// New2 constructs a new QHostAddress object.
    pub fn New2(ip4Addr: u32) QtC.QHostAddress {
        return qtc.QHostAddress_new2(ip4Addr);
    }


    /// New3 constructs a new QHostAddress object.
    pub fn New3(ip6Addr: *const u8) QtC.QHostAddress {
        return qtc.QHostAddress_new3(@ptrCast(ip6Addr));
    }


    /// New4 constructs a new QHostAddress object.
    pub fn New4(ip6Addr: ?*anyopaque) QtC.QHostAddress {
        return qtc.QHostAddress_new4(@ptrCast(ip6Addr));
    }


    /// New5 constructs a new QHostAddress object.
    pub fn New5(address: []const u8) QtC.QHostAddress {
        const address_str = qtc.libqt_string{
    .len = address.len,
    .data = address.ptr,
};
return qtc.QHostAddress_new5(address_str);
    }


    /// New6 constructs a new QHostAddress object.
    pub fn New6(copyVal: ?*anyopaque) QtC.QHostAddress {
        return qtc.QHostAddress_new6(@ptrCast(copyVal));
    }


    /// New7 constructs a new QHostAddress object.
    pub fn New7(address: i32) QtC.QHostAddress {
        return qtc.QHostAddress_new7(@intCast(address));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QHostAddress_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorAssign2(self: ?*anyopaque, address: i32) void {
        qtc.QHostAddress_OperatorAssign2(@ptrCast(self), @intCast(address));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QHostAddress_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn SetAddress(self: ?*anyopaque, ip4Addr: u32) void {
        qtc.QHostAddress_SetAddress(@ptrCast(self), ip4Addr);
    }

    pub fn SetAddress2(self: ?*anyopaque, ip6Addr: *const u8) void {
        qtc.QHostAddress_SetAddress2(@ptrCast(self), @ptrCast(ip6Addr));
    }

    pub fn SetAddress3(self: ?*anyopaque, ip6Addr: ?*anyopaque) void {
        qtc.QHostAddress_SetAddress3(@ptrCast(self), @ptrCast(ip6Addr));
    }

    pub fn SetAddress5(self: ?*anyopaque, address: []const u8) bool {
        const address_str = qtc.libqt_string{
    .len = address.len,
    .data = address.ptr,
};
return qtc.QHostAddress_SetAddress5(@ptrCast(self), address_str);
    }

    pub fn SetAddress6(self: ?*anyopaque, address: i32) void {
        qtc.QHostAddress_SetAddress6(@ptrCast(self), @intCast(address));
    }

    pub fn Protocol(self: ?*anyopaque, ) i32 {
        return qtc.QHostAddress_Protocol(@ptrCast(self));
    }

    pub fn ToIPv4Address(self: ?*anyopaque, ) u32 {
        return qtc.QHostAddress_ToIPv4Address(@ptrCast(self));
    }

    pub fn ToIPv6Address(self: ?*anyopaque, ) QtC.QIPv6Address {
        return qtc.QHostAddress_ToIPv6Address(@ptrCast(self));
    }

    pub fn ToString(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QHostAddress_ToString(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qhostaddress.ToString: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn ScopeId(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QHostAddress_ScopeId(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qhostaddress.ScopeId: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetScopeId(self: ?*anyopaque, id: []const u8) void {
        const id_str = qtc.libqt_string{
    .len = id.len,
    .data = id.ptr,
};
qtc.QHostAddress_SetScopeId(@ptrCast(self), id_str);
    }

    pub fn IsEqual(self: ?*anyopaque, address: ?*anyopaque) bool {
        return qtc.QHostAddress_IsEqual(@ptrCast(self), @ptrCast(address));
    }

    pub fn OperatorEqual(self: ?*anyopaque, address: ?*anyopaque) bool {
        return qtc.QHostAddress_OperatorEqual(@ptrCast(self), @ptrCast(address));
    }

    pub fn OperatorEqual2(self: ?*anyopaque, address: i32) bool {
        return qtc.QHostAddress_OperatorEqual2(@ptrCast(self), @intCast(address));
    }

    pub fn OperatorNotEqual(self: ?*anyopaque, address: ?*anyopaque) bool {
        return qtc.QHostAddress_OperatorNotEqual(@ptrCast(self), @ptrCast(address));
    }

    pub fn OperatorNotEqual2(self: ?*anyopaque, address: i32) bool {
        return qtc.QHostAddress_OperatorNotEqual2(@ptrCast(self), @intCast(address));
    }

    pub fn IsNull(self: ?*anyopaque, ) bool {
        return qtc.QHostAddress_IsNull(@ptrCast(self));
    }

    pub fn Clear(self: ?*anyopaque, ) void {
        qtc.QHostAddress_Clear(@ptrCast(self));
    }

    pub fn IsInSubnet(self: ?*anyopaque, subnet: ?*anyopaque, netmask: i32) bool {
        return qtc.QHostAddress_IsInSubnet(@ptrCast(self), @ptrCast(subnet), netmask);
    }

    pub fn IsLoopback(self: ?*anyopaque, ) bool {
        return qtc.QHostAddress_IsLoopback(@ptrCast(self));
    }

    pub fn IsGlobal(self: ?*anyopaque, ) bool {
        return qtc.QHostAddress_IsGlobal(@ptrCast(self));
    }

    pub fn IsLinkLocal(self: ?*anyopaque, ) bool {
        return qtc.QHostAddress_IsLinkLocal(@ptrCast(self));
    }

    pub fn IsSiteLocal(self: ?*anyopaque, ) bool {
        return qtc.QHostAddress_IsSiteLocal(@ptrCast(self));
    }

    pub fn IsUniqueLocalUnicast(self: ?*anyopaque, ) bool {
        return qtc.QHostAddress_IsUniqueLocalUnicast(@ptrCast(self));
    }

    pub fn IsMulticast(self: ?*anyopaque, ) bool {
        return qtc.QHostAddress_IsMulticast(@ptrCast(self));
    }

    pub fn IsBroadcast(self: ?*anyopaque, ) bool {
        return qtc.QHostAddress_IsBroadcast(@ptrCast(self));
    }

    pub fn IsPrivateUse(self: ?*anyopaque, ) bool {
        return qtc.QHostAddress_IsPrivateUse(@ptrCast(self));
    }

    pub fn ParseSubnet(subnet: []const u8) QPair<QHostAddress, int> {
        const subnet_str = qtc.libqt_string{
    .len = subnet.len,
    .data = subnet.ptr,
};
return qtc.QHostAddress_ParseSubnet(subnet_str);
    }

    pub fn ToIPv4Address1(self: ?*anyopaque, ok: *bool) u32 {
        return qtc.QHostAddress_ToIPv4Address1(@ptrCast(self), @ptrCast(ok));
    }

    pub fn IsEqual2(self: ?*anyopaque, address: ?*anyopaque, mode: i32) bool {
        return qtc.QHostAddress_IsEqual2(@ptrCast(self), @ptrCast(address), @intCast(mode));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QHostAddress_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/hostaddress.html#types
pub const enums = struct {
    pub const SpecialAddress = enum {
        pub const Null: i32 = 0;
        pub const Broadcast: i32 = 1;
        pub const LocalHost: i32 = 2;
        pub const LocalHostIPv6: i32 = 3;
        pub const Any: i32 = 4;
        pub const AnyIPv6: i32 = 5;
        pub const AnyIPv4: i32 = 6;
    };

    pub const ConversionModeFlag = enum {
        pub const ConvertV4MappedToIPv4: i32 = 1;
        pub const ConvertV4CompatToIPv4: i32 = 2;
        pub const ConvertUnspecifiedAddress: i32 = 4;
        pub const ConvertLocalHost: i32 = 8;
        pub const TolerantConversion: i32 = 255;
        pub const StrictConversion: i32 = 0;
    };

};
